modules/siddhi-extensions/markov-models/src/main/java/org/wso2/siddhi/extension/markov/MarkovChainTransitionProbabilitiesCalculator.java

```diff
@@ -0,0 +1,237 @@
+    }
+
+    /**
+     * Method to process new events
+     *
+     * @param userId
+     * @param state
+     * @param trainingOption
+     * @return Array containing output results
+     */
+    protected Object[] processData(String userId, String state, boolean trainingOption) {
+
+        eventCount++;
+        String id = userId;
+        String currentState = state;
+        boolean currentTrainingOption = trainingOption;
+        Object[] markovChainAlert = null;
+
+        String lastState = updateStates(id, currentState);
+
+        if (eventCount > notificationsHoldLimit) {
+            markovChainAlert = getMarkovChainAlert(lastState, currentState);
+        }
+
+        if (currentTrainingOption && lastState != null) {
+            markovMatrix.updateStartStateCount(lastState, 1);
+            markovMatrix.updateTransitionCount(lastState, currentState, 1);
+        }
+
+        return markovChainAlert;
+    }
+
+    /**
+     * Method to updateTransitionCount States
+     */
+    private String updateStates(String id, String currentState) {
+
+        String lastState = idToLastStates.get(id);
+        // updateTransitionCount the last state to current state
+        idToLastStates.put(id, currentState);
+
+        // updating expiry times
+        long currentEventTime = System.currentTimeMillis();
+        long expiryTime = currentEventTime + durationToKeep;
+        idToLastStatesExpiryTime.put(id, expiryTime);
+
+        return lastState;
+    }
+
+    /**
+     * Method which updateTransitionCount markov matrix and return notifications
+     *
+     * @return Array containing last state, transition probability and notify values
+     */
+    private Object[] getMarkovChainAlert(String lastState, String currentState) {
+
+        Double stateTransitionProbability = 0.0;
+        String key;
+        Double totalCount;
+        Double transitionCount;
+        boolean notify = false;
+
+        if (lastState != null) {
+            key = markovMatrix.getKey(lastState, currentState);
+
+            totalCount = markovMatrix.getStartStateCount().get(lastState);
+            transitionCount = markovMatrix.getTransitionCount().get(key);
+            if (totalCount != null && transitionCount != null) {
+                stateTransitionProbability = transitionCount / totalCount;
+            }
+
+            if (stateTransitionProbability <= alertThresholdProbability) {
+                notify = true;
+            }
+        }
+        return new Object[] { lastState, stateTransitionProbability, notify };
+    }
+
+    /**
+     * Method to remove events which have expired based on time
+     */
+    protected void removeExpiredEvents(long currentEventTime) {
+
+        for (Map.Entry<String, Long> entry : idToLastStatesExpiryTime.entrySet()) {
+            if (entry.getValue() <= currentEventTime) {
+                idToLastStates.remove(entry.getKey());
+            } else {
+                break;
```

Comment:
Since idToLastStatesExpiryTime is a LinkedHashMap when we iterating it, elements will be returned in the order in which they were inserted. Therefore if a particular value is not less than or equal to curcurrentEventTime the next values also won't satisfy the condition. Therefore we don't need to iterate trough the whole map.